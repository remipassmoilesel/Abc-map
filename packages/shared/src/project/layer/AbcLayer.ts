/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { FeatureCollection } from 'geojson';
import { AbcProjection } from '../projection/AbcProjection';
import { LayerAuthentication } from './LayerAuthentication';

export type AbcLayer = AbcVectorLayer | AbcPredefinedLayer | AbcWmsLayer | AbcWmtsLayer | AbcXyzLayer;

export type LayerMetadata = VectorMetadata | PredefinedMetadata | WmsMetadata | WmtsMetadata | XyzMetadata;

export enum LayerType {
  Vector = 'Vector',
  Predefined = 'Predefined',
  Wms = 'Wms',
  Wmts = 'Wmts',
  Xyz = 'Xyz',
}

// Base types

export interface AbcBaseLayer {
  type: LayerType;
  metadata: LayerMetadata;
}

export interface BaseMetadata {
  id: string;
  name: string;
  opacity: number;
  visible: boolean;
  active: boolean;
  type: LayerType;
  /**
   * Users can add a custom attribution to a layer (for exports).
   * But they can not modify existing ones (like OSM attributions).
   */
  attributions?: string[];
}

// Vector layer

export interface VectorMetadata extends BaseMetadata {
  type: LayerType.Vector;
}

export interface AbcVectorLayer extends AbcBaseLayer {
  type: LayerType.Vector;
  metadata: VectorMetadata;
  features: FeatureCollection;
}

// Predefined layers

export interface PredefinedMetadata extends BaseMetadata {
  type: LayerType.Predefined;
  model: PredefinedLayerModel;
}

export interface AbcPredefinedLayer extends AbcBaseLayer {
  type: LayerType.Predefined;
  metadata: PredefinedMetadata;
}

export enum PredefinedLayerModel {
  OSM = 'OSM',
  StamenToner = 'StamenToner',
  StamenTonerLite = 'StamenTonerLite',
  StamenTerrain = 'StamenTerrain',
  StamenWatercolor = 'StamenWatercolor',
}

// WMS

export interface WmsMetadata extends BaseMetadata {
  type: LayerType.Wms;
  projection?: AbcProjection;
  extent?: [number, number, number, number];
  remoteUrls: string[];
  remoteLayerName: string;
  auth?: LayerAuthentication;
}

export interface AbcWmsLayer extends AbcBaseLayer {
  type: LayerType.Wms;
  metadata: WmsMetadata;
}

// WMTS

export interface WmtsMetadata extends BaseMetadata {
  type: LayerType.Wmts;
  capabilitiesUrl: string;
  remoteLayerName: string;
  auth?: LayerAuthentication;
}

export interface AbcWmtsLayer extends AbcBaseLayer {
  type: LayerType.Wmts;
  metadata: WmtsMetadata;
}

// XYZ

export interface XyzMetadata extends BaseMetadata {
  type: LayerType.Xyz;
  projection?: AbcProjection;
  remoteUrl: string;
}

export interface AbcXyzLayer extends AbcBaseLayer {
  type: LayerType.Xyz;
  metadata: XyzMetadata;
}
