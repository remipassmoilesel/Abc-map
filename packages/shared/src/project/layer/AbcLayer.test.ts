/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcPredefinedLayer, AbcVectorLayer, AbcWmsLayer, AbcXyzLayer, LayerType, PredefinedLayerModel } from './AbcLayer';

/**
 * If this test fail, you should write a migration script then adapt it
 */
describe('AbcLayer', () => {
  it('VectorLayer should not change without migration', () => {
    const current: AbcVectorLayer = {
      type: LayerType.Vector,
      metadata: {
        id: 'test-layer-id',
        name: 'Test vector layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Vector,
      },
      features: {
        type: 'FeatureCollection',
        bbox: [1, 2, 3, 4],
        features: [
          {
            id: 'feature-id',
            type: 'Feature',
            bbox: [5, 6, 7, 8],
            properties: {
              property1: 'value1',
            },
            geometry: {
              type: 'Point',
              coordinates: [9, 10],
            },
          },
        ],
      },
    };

    expect(current).toMatchSnapshot();
  });

  it('PredefinedLayer should not change without migration', () => {
    const currentLayer: AbcPredefinedLayer = {
      type: LayerType.Predefined,
      metadata: {
        id: 'test-layer-id',
        name: 'Test predefined layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Predefined,
        model: PredefinedLayerModel.OSM,
      },
    };

    expect(currentLayer).toMatchSnapshot();
    expect(PredefinedLayerModel).toMatchSnapshot();
  });

  it('WMS layer should not change without migration', () => {
    const currentLayer: AbcWmsLayer = {
      type: LayerType.Wms,
      metadata: {
        id: 'test-layer-id',
        name: 'Test wms layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Wms,
        auth: {
          username: 'test-username',
          password: 'test-password',
        },
        extent: [1, 1, 1, 1],
        projection: { name: 'EPSG:4326' },
        remoteLayerName: 'test-remoteLayerName',
        remoteUrls: ['test-remoteUrls'],
      },
    };

    expect(currentLayer).toMatchSnapshot();
  });

  it('XYZ layer should not change without migration', () => {
    const currentLayer: AbcXyzLayer = {
      type: LayerType.Xyz,
      metadata: {
        id: 'test-layer-id',
        name: 'Test wms layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Xyz,
        remoteUrl: 'test-remoteUrl',
      },
    };

    expect(currentLayer).toMatchSnapshot();
  });
});
