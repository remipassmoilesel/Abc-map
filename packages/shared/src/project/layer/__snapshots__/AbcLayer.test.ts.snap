// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbcLayer PredefinedLayer should not change without migration 1`] = `
Object {
  "metadata": Object {
    "active": true,
    "id": "test-layer-id",
    "model": "OSM",
    "name": "Test predefined layer",
    "opacity": 0.5,
    "type": "Predefined",
    "visible": true,
  },
  "type": "Predefined",
}
`;

exports[`AbcLayer PredefinedLayer should not change without migration 2`] = `
Object {
  "OSM": "OSM",
}
`;

exports[`AbcLayer VectorLayer should not change without migration 1`] = `
Object {
  "features": Object {
    "bbox": Array [
      1,
      2,
      3,
      4,
    ],
    "features": Array [
      Object {
        "bbox": Array [
          5,
          6,
          7,
          8,
        ],
        "geometry": Object {
          "coordinates": Array [
            9,
            10,
          ],
          "type": "Point",
        },
        "id": "feature-id",
        "properties": Object {
          "property1": "value1",
        },
        "type": "Feature",
      },
    ],
    "type": "FeatureCollection",
  },
  "metadata": Object {
    "active": true,
    "id": "test-layer-id",
    "name": "Test vector layer",
    "opacity": 0.5,
    "type": "Vector",
    "visible": true,
  },
  "type": "Vector",
}
`;

exports[`AbcLayer WMS layer should not change without migration 1`] = `
Object {
  "metadata": Object {
    "active": true,
    "auth": Object {
      "password": "test-password",
      "username": "test-username",
    },
    "extent": Array [
      1,
      1,
      1,
      1,
    ],
    "id": "test-layer-id",
    "name": "Test wms layer",
    "opacity": 0.5,
    "projection": Object {
      "name": "EPSG:4326",
    },
    "remoteLayerName": "test-remoteLayerName",
    "remoteUrls": Array [
      "test-remoteUrls",
    ],
    "type": "Wms",
    "visible": true,
  },
  "type": "Wms",
}
`;

exports[`AbcLayer XYZ layer should not change without migration 1`] = `
Object {
  "metadata": Object {
    "active": true,
    "id": "test-layer-id",
    "name": "Test wms layer",
    "opacity": 0.5,
    "remoteUrl": "test-remoteUrl",
    "type": "Xyz",
    "visible": true,
  },
  "type": "Xyz",
}
`;
